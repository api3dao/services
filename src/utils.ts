// TODO: Finalize and publish this package on npm and import it here
export type GoResultSuccess<T> = [null, T];
export type GoResultError<E = Error> = [E, null];
export type GoResult<T, E = Error> = GoResultSuccess<T> | GoResultError<E>;
export const GO_ERROR_INDEX = 0;
export const GO_RESULT_INDEX = 1;

export const successFn = <T>(value: T): [null, T] => [null, value];
export const errorFn = <E = Error>(err: E): [E, null] => [err, null];

export const goSync = <T>(fn: () => T): GoResult<T> => {
  try {
    return successFn(fn());
  } catch (err) {
    return errorFn(err as Error);
  }
};

export const go = async <T>(fn: Promise<T> | (() => Promise<T>)): Promise<GoResult<T>> => {
  // We need try/catch because `fn` might throw sync errors as well
  try {
    if (typeof fn === 'function') {
      return fn().then(successFn).catch(errorFn);
    }
    return fn.then(successFn).catch(errorFn);
  } catch (err) {
    return errorFn(err as Error);
  }
};

export const isGoSuccess = <T, E>(result: GoResult<T, E>): result is GoResultSuccess<T> => !result[GO_ERROR_INDEX];

export const rethrowError = (error: Error) => {
  throw error;
};

// NOTE: This needs to be written using 'function' syntax (cannot be arrow function)
// See: https://github.com/microsoft/TypeScript/issues/34523#issuecomment-542978853
export function assertGoSuccess<T>(result: GoResult<T>, onError = rethrowError): asserts result is GoResultSuccess<T> {
  if (result[0]) {
    onError(result[0]);
  }
}
