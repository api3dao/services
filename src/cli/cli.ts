import { mkdirSync, readdirSync } from 'fs';
import { join } from 'path';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import prompts, { PromptObject } from 'prompts';
import { copySync } from 'fs-extra';
import chalk from 'chalk';
import { go, isGoSuccess } from '../utils';

const promptQuestions = (questions: PromptObject[]) =>
  prompts(questions, {
    // https://github.com/terkelg/prompts/issues/27#issuecomment-527693302
    onCancel: () => {
      throw new Error('Aborted by the user');
    },
  });

const createCliOption = (name: string) => ({
  title: name,
  value: name,
});

const TEMPLATE_DIR = join(__dirname, '../../templates');
export const getAvailableTemplates = () => readdirSync(TEMPLATE_DIR);
export const getTemplatePath = (template: string) => join(TEMPLATE_DIR, template);

// TODO: test
export const createProjectUsingTemplate = (path: string, template: string) => {
  console.info(`Creating a new beacon reader app in ${chalk.green(path)}`);

  mkdirSync(path, { recursive: true });
  copySync(getTemplatePath(template), path);

  // TODO: install dependencies

  console.info(`Successfully created and initialized a new project.`);
};

export const createCli = () =>
  yargs(hideBin(process.argv))
    .command(
      // The name of the cli binary is "create-beacon-reader-app" and it should generate the without the need to specify a
      // command. See: https://github.com/yargs/yargs/blob/main/docs/advanced.md#default-commands
      '$0',
      'Create a project skeleton for an application that reads a value from beacon',
      {
        path: {
          description:
            "Path to a directory in which to create a project. The directory will be created if it doesn't exist yet.",
          demandOption: false,
          type: 'string',
        },
        template: {
          description:
            'The name of the template to be used. Use the "show-available-templates" command to list the available options.',
          demandOption: false,
          type: 'string',
        },
      },
      // TODO: Test interactivity
      async (args) => {
        // We validate the template first, because we want to show the output of "--help" command before we prompt the
        // user for questions. We do this because the mixed output of "--help" and prompts doesn't look good.
        const templates = getAvailableTemplates();
        if (args.template && !templates.includes(args.template)) throw new Error(`Unknown template: ${args.template}`);

        const goResult = await go(
          promptQuestions([
            {
              type: args.path ? null : 'text',
              name: 'path',
              message: [
                'Please, specify a path to the directory in which to create the project.',
                "(The directory will be created if it doesn't exist)",
              ].join('\n'),
            },
            {
              type: args.template ? null : 'select',
              name: 'template',
              message: 'Please, pick one of the available templates from the list',
              choices: templates.map(createCliOption),
            },
          ])
        );

        // An error here means the user cancelled the prompt question (e.g. ctrl+c).
        if (!isGoSuccess(goResult)) return;
        // We need to spread the parameters from "yargs" since if a user defined "path" as argument it will not be part
        // of the result from question prompt.
        const { path, template } = { ...args, ...goResult.data } as Record<string, string>;

        createProjectUsingTemplate(path, template);
      }
    )
    .command('show-available-templates', 'Show the list of available template options', {}, () => {
      console.info(getAvailableTemplates().join('\n'));
    })
    .example([
      ['$0', 'Run the interactive version of the CLI'],
      ['$0 --path="."', 'Create the project in the current directory'],
      [
        `$0 --path="my-app" --template="${getAvailableTemplates()[0]}"`,
        'Create the project inside the "my-app" directory',
      ],
    ])
    .help()
    .recommendCommands()
    .strict()
    .wrap(120).argv;
