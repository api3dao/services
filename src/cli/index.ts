#!/usr/bin/env node
import { createCli } from './cli';

createCli();
