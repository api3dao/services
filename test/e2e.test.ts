import { ChildProcessWithoutNullStreams, execSync, spawn } from 'child_process';
import { chmodSync } from 'fs';
import chalk from 'chalk';
import { createTmpDir } from './utils';
import { getAvailableTemplates } from '../src/cli/cli';

const CLI_EXECUTABLE = `${__dirname}/../dist/cli/index.js`;
// Turning this flag to 'true' will print each command before executing it
// It might be useful to turn on, while debugging particular test.
const DEBUG_COMMANDS = false;

it('has disabled DEBUG_COMMANDS flag', () => {
  expect(DEBUG_COMMANDS).toBe(false);
});

type CommandArg = [string, string | string[] | number | boolean];
const execCommand = (command: string, ...args: CommandArg[]) => {
  const quote = (val: string) => `"${val}"`;
  const formattedArgs = args
    .map(([c, a]) => {
      // if args is array then quote each elem and separate them with space
      if (Array.isArray(a)) return `${c} ${a.map(quote).join(' ')}`;
      // otherwise just quote each elem and separate them with space
      else return `${c} ${quote(String(a))}`;
    })
    .join(' ');
  const formattedCommand = `${command} ${formattedArgs}`;
  if (DEBUG_COMMANDS) console.log(`Executing command: ${formattedCommand}`);
  try {
    return execSync(`${CLI_EXECUTABLE} ${formattedCommand}`).toString().trim();
  } catch (e: any) {
    // rethrow the output of the CLI
    throw new Error(e.stdout.toString().trim());
  }
};

describe('Beacon reader CLI', () => {
  beforeAll(() => {
    chmodSync(CLI_EXECUTABLE, '755');
  });

  it('shows help', () => {
    const output = execCommand('--help');
    expect(output).toMatchSnapshot();
  });

  it('shows available commands', () => {
    const output = execCommand('show-available-templates');
    expect(output).toEqual(['javascript-ethers'].join('\n'));
  });

  describe('interactive command', () => {
    const waitOnOutput = (childProcess: ChildProcessWithoutNullStreams) =>
      new Promise<string>((resolve) =>
        childProcess.stdout.once('readable', () => {
          const data = childProcess.stdout.read().toString();
          resolve(data);
        })
      );

    it('works', async () => {
      // Run the command in a child process so we can pipe stdin to it
      const childProcess = spawn(CLI_EXECUTABLE);

      // Write the path to the script
      const tmpDir = createTmpDir();
      childProcess.stdin.write(tmpDir + '\n');
      await waitOnOutput(childProcess);

      // Choose the template to create (accept the default one)
      childProcess.stdin.write('\n');
      // Wait a few seconds until the script initializes the project
      await new Promise((res) => setTimeout(res, 5000));
      const output = await waitOnOutput(childProcess);

      expect(childProcess.kill()).toBe(true);
      expect(output).toContain(`Creating a new beacon reader app in ${chalk.green(tmpDir)}`);
      expect(output).toContain(`Successfully created and initialized a new project.`);
    }, 10_000);
  });

  it('parses command arguments', () => {
    const tmpDir = createTmpDir();
    const template = 'javascript-ethers';

    const output = execCommand('', ['--path', tmpDir], ['--template', template]);

    const lines = output.split('\n');
    expect(lines[0]).toContain(`Creating a new beacon reader app in ${chalk.green(tmpDir)}`);
    expect(lines[1]).toContain(`Successfully created and initialized a new project.`);
  });

  describe('templates are functional', () => {
    getAvailableTemplates().forEach((templateName) => {
      it(`tests ${templateName}`, () => {
        const tmpDir = createTmpDir();

        // 1. Create a repository based on the current template name
        execCommand('', ['--path', tmpDir], ['--template', templateName]);

        // 2. Install the dependencies for the created project
        process.chdir(tmpDir);
        // TODO: Is it OK to be opinionated about yarn or shall we also have npm templates?
        // Redirect the stderr of yarn install to /dev/null to avoid the output appearing in jest output
        const installOutput = execSync(`FORCE_COLOR=0 yarn install 2>/dev/null`).toString();
        expect(installOutput).toContain('success Saved lockfile.');

        // 3. Run the sample beacon reader tests
        // TODO: This should be a nicer script inside the package.json of the template
        const testOutput = execSync(`FORCE_COLOR=0 npx hardhat test test/BeaconReaderExample.test.js`).toString();
        expect(testOutput).toContain('1 passing');

        // 4. Deploy the beacon reader (and mocked beacon server) and read a value from it
        // TODO: This should be a nicer script inside the package.json of the template
        const scriptOutput = execSync(
          `FORCE_COLOR=0 npx hardhat run scripts/deploy.js && npx hardhat run scripts/read-beacon.js`
        ).toString();
        expect(scriptOutput).toContain('BeaconReaderExample deployed to:');
        expect(scriptOutput).toContain('Beacon value:');

        // 5. TODO: Read a value from a beacon that is already deployed. (use services repo for this)
      });
    });
  });
});
