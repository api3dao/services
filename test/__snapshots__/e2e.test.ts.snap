// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beacon reader CLI shows help 1`] = `
"index.js

Create a project skeleton for an application that reads a value from beacon

Commands:
  index.js                           Create a project skeleton for an application that reads a value from beacon
                                                                                                               [default]
  index.js show-available-templates  Show the list of available template options

Options:
  --version   Show version number                                                                              [boolean]
  --help      Show help                                                                                        [boolean]
  --path      Path to a directory in which to create a project. The directory will be created if it doesn't exist yet.
                                                                                                                [string]
  --template  The name of the template to be used. Use the \\"show-available-templates\\" command to list the available
              options.                                                                                          [string]

Examples:
  index.js                                                 Run the interactive version of the CLI
  index.js --path=\\".\\"                                      Create the project in the current directory
  index.js --path=\\"my-app\\" --template=\\"javascript-ethers\\"  Create the project inside the \\"my-app\\" directory"
`;
